### PR TITLE
[TF:TRT] Replace c asserts with CHECK macro

### PR DESCRIPTION
**Pull request by christopherbate**
The default Tensorflow build uses `-c opt`, which implies `-NDEBUG`. Thus, c-style assert statements become no-ops. This PR replaces asserts in ITensorProxy with Tensorflow CHECK macro. This induces failures in the tests, which are corrected.

Test failures indicate no one is using the `-c debug` option, and thus the signal given by the assert statements is being missed. Thus they should be changed to CHECKs.

@bixia1 
@jhalakpatel 
@tfeher

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
Please use
 - `CHECK_EQ(X, Y)` instead of `CHECK(X == Y)`
 - `LOG(FATAL) << msg` instead of `CHECK(0 && msg)`

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 1]
```diff
@@ -767,8 +767,8 @@ TEST_F(ConverterTest, AddAndGetInputs) {
   TF_EXPECT_OK(GetInputs(node_def, &inputs));
 
   EXPECT_EQ(4, inputs.size());
-  EXPECT_EQ(inputs[0].tensor()->simple_tensor(),
-            inputs[1].tensor()->simple_tensor());
+  EXPECT_EQ(inputs[0].tensor()->trt_tensor(),
```
Comment (sanjoy) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:770]: These seem unrelated to the PR description.

**Comment by bixia1** [comment]
Within google, there is a meta data file that prevents us from using CHECK in TF-TRT code. For this reason, your change to trt_tensor_proxy.h  won't be accepted. You might still see some uses of CHECK in TF-TRT, which are added before we have that policy. 

Are you also trying to fix a bug in the test? Would you please revise your PR for this?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 1 above)
Comment (christopherbate) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:770]: It's not. These hit the assert statement but since the assert was a noop, we didn't see the effect. When we change to CHECK these throw errors and we need to change to "trt_tensor()"

**Comment by christopherbate** [comment]
>  a meta data file that prevents us from using CHECK in TF-TRT code. For this reason, your change to trt_tensor_proxy.h won't be accepted. The meta data is not in the open source though, but the reason behind this policy is to prevent TensorFlow from crashing for any reason.
> 
> You might still see some uses of CHECK in TF-TRT, which are added before we have that policy.

So are you running CI/CD against some sort of debug profile to see what the asserts do? Because the tests are obviously failing when the asserts are not turned off by `-NDEBUG` flag which is present by default. That's what this PR is addressing. If you have a policy against CHECK then you should have a policy against `assert` as well?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 1 above)
Comment (christopherbate) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:770]: @bixia1 this is one place where the tests are incorrect (an assert would be triggered in debug build), but under default profile it doesn't complain. There are a couple other places I fixed as well.

**Comment by christopherbate** [comment]
We should just have a `CHECK` wrapper that you can turn off via a compile time flag and we (OSS) can leave on?

**Comment by bixia1** [comment]
The policy is to prevent having CHECK in the source code, I tried that even
  if (0) CHECK(...)
can't be accepted.

On the other hand, can you build with debug for internal check?

Since we are talking about checking the internal logic in TF-TRT, do you see any problem in the code that can make these check fail? We should fix the internal logic in TF-TRT instead.

**Comment by christopherbate** [comment]
I can catch things only when the asserts are not disabled. I've been trying to add a profile for us internally to run just TF-TRT with ASAN and debug on, but I haven't gotten that to work yet.

As far as internal logic, I fixed  the things that I found by running the unit tests. However, that doesn't rule out the existence of a runtime bug which is not covered by unit tests. 

As far as code structure, the `ITensorProxyPtr` will need to be "redesigned". The funky fake runtime polymorphism used there is causing all the problems and increases testing burden (hence all the `asserts` in the code there).

**Comment by bixia1** [comment]
You can introduce something like TF_TRT_ASSERT, and make it equivalent to assert in the tree. Then you only need to change this macro in your local workspace for your purpose. Will this be good enough?

**Comment by mihaimaruseac** [comment]
I'm working on getting that policy changed. Will take a while though but we've had security issues where existing `CHECK`s in the past have been just removed / been replaced with assert-like statements only active in debug builds.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 1 above)
Comment (sanjoy) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:770]: SG.

**Comment by bixia1** [comment]
@christopherbate Please hold on to this, I am checking to see if we can allow CHECK in this PR and will update here once I get a conclusion.

File: tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h [hunk 2]
```diff
@@ -143,28 +143,26 @@ class ITensorProxy {
...
+    CHECK(validate());
     return simple_tensor_ != nullptr;
   }
 
   TensorType ttype() const { return ttype_; }
 
   nvinfer1::ITensor* trt_tensor() const {
-    assert(trt_tensor_ != nullptr);
-    assert(ttype_ == TensorType::kTRT);
+    CHECK(trt_tensor_ != nullptr);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:158]: CHECK_NOTNULL

File: tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h [hunk 3]
```diff
@@ -143,28 +143,26 @@ class ITensorProxy {
...
+    CHECK(ttype_ == TensorType::kTRT);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:159]: CHECK_EQ probably doesn't work here because of the missing of an output operator for TensorType enum.
We can keep this or add an output operator. I am fine with either way.

File: tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h [hunk 4]
```diff
@@ -143,28 +143,26 @@ class ITensorProxy {
...
     return trt_tensor_;
   }
 
   SimpleITensor* simple_tensor() const {
-    assert(simple_tensor_ != nullptr);
-    assert(ttype_ == TensorType::kSIMPLE);
+    CHECK(simple_tensor_ != nullptr);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:164]: CHECK_NOTNULL

File: tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h [hunk 5]
```diff
@@ -175,7 +173,7 @@ class ITensorProxy {
       case TensorType::kSIMPLE:
         return simple_tensor_->setName(name);
     }
-    assert(0 && "Unsupported itensor_ type");
+    CHECK(0 && "Unsupported itensor_ type");
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h:176]: LOG(FATAL)  << followed by the message and the value for is used.
Similar comment for all the CHECK(0& below

**Comment by christopherbate** [comment]
Comments addressed.

**Review by bixia1 — APPROVED** [review]
